elastic/kibana
# [ES UI Shared] Monaco XJSON

## Summary

Add XJSON mode support for Monaco. The intention is that this be used inside of Watcher, ML and any other new editors that we build so that we are able to use Monaco instead of Ace.

## How to test

1. Start Kibana
2. Go to Painless Lab (this has been hacked to be an XJSON mode editor temporarily)
3. Paste the following XJSON into the editor:

```
{
  "test": """test""",
  "script": """ while if else true """,
  "test": "sweet!",
  "else": true,
  "canBe": false,
  "object": {
    "inside": {
      "is": "another"
    }
  }
}
```

The XJSON should be highlighted correctly and indicate that there is a duplicate key "test".

## Owners of specific sections

- Note to ops, created a new package called `@kbn/langs`. In future I would like to put the grammar parsers for Console here too. This is similar to the `@kbn/interpreter` folder which is apparently going to be removed before (8.0.0) CC (@ppisljar). The alternative is that we put this inside es-ui-shared too. Any alternate suggestions welcome of course. I was thinking we can also put SQL, EQL and any other parsers we build here (if we build them).
- Note to ML (@darnautov) please test the painless lab as described above, and test that your previous use of `useXJsonMode` works as expected


## TODO

- [x] Implement an analogous XJSON mode hook for use with monaco.
- [x] Revert the dummy Painless Lab integration


## Screenshots

<img width="470" alt="Screenshot 2020-05-27 at 16 07 42" src="https://user-images.githubusercontent.com/8155004/83032353-01607c00-a036-11ea-988c-d8a79ca702dd.png">

### Additional Context

See this issue: https://github.com/elastic/kibana/issues/57435
